# Pré Processamento de Dados

## Import das Bibliotecas

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import html

## Carregamento dos Dados

In [ ]:
df_books_v2 = pd.read_csv('Books.csv')
df_ratings_v2 = pd.read_csv('Ratings.csv')
df_users_v2 = pd.read_csv('Users.csv')

<ipython-input-2-ebc4332a0ca1>:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  df_books_v2 = pd.read_csv('Books.csv')


## Limpeza e Tratamento nos Dados

##**`df_books`**

### Valores Ausentes

In [ ]:
df_books_v2.set_index('ISBN', inplace=True)

In [ ]:
df_books_v2[df_books_v2.isnull().any(axis=1)]

Book-Title  \
ISBN                                                            
0751352497                           A+ Quiz Masters:01 Earth   
193169656X                                        Tyrant Moon   
1931696993                                    Finders Keepers   
9627982032  The Credit Suisse Guide to Managing Your Perso...   
078946697X  DK Readers: Creating the X-Men, How It All Beg...   
2070426769  Peuple du ciel, suivi de 'Les Bergers\";Jean-M...   
0789466953  DK Readers: Creating the X-Men, How Comic Book...   

                Book-Author Year-Of-Publication  \
ISBN                                              
0751352497              NaN                1999   
193169656X  Elaine Corvidae                2002   
1931696993  Linnea Sinclair                2001   
9627982032              NaN                1995   
078946697X             2000   DK Publishing Inc   
2070426769             2003           Gallimard   
0789466953             2000   DK Publishing Inc   

                                                    Publisher  \
ISBN                                                            
0751352497                                 Dorling Kindersley   
193169656X                                                NaN   
1931696993                                                NaN   
9627982032                     Edinburgh Financial Publishing   
078946697X  http://images.amazon.com/images/P/078946697X.0...   
2070426769  http://images.amazon.com/images/P/2070426769.0...   
0789466953  http://images.amazon.com/images/P/0789466953.0...   

                                                  Image-URL-S  \
ISBN                                                            
0751352497  http://images.amazon.com/images/P/0751352497.0...   
193169656X  http://images.amazon.com/images/P/193169656X.0...   
1931696993  http://images.amazon.com/images/P/1931696993.0...   
9627982032  http://images.amazon.com/images/P/9627982032.0...   
078946697X  http://images.amazon.com/images/P/078946697X.0...   
2070426769  http://images.amazon.com/images/P/2070426769.0...   
0789466953  http://images.amazon.com/images/P/0789466953.0...   

                                                  Image-URL-M  \
ISBN                                                            
0751352497  http://images.amazon.com/images/P/0751352497.0...   
193169656X  http://images.amazon.com/images/P/193169656X.0...   
1931696993  http://images.amazon.com/images/P/1931696993.0...   
9627982032  http://images.amazon.com/images/P/9627982032.0...   
078946697X  http://images.amazon.com/images/P/078946697X.0...   
2070426769  http://images.amazon.com/images/P/2070426769.0...   
0789466953  http://images.amazon.com/images/P/0789466953.0...   

                                                  Image-URL-L  
ISBN                                                           
0751352497  http://images.amazon.com/images/P/0751352497.0...  
193169656X  http://images.amazon.com/images/P/193169656X.0...  
1931696993  http://images.amazon.com/images/P/1931696993.0...  
9627982032  http://images.amazon.com/images/P/9627982032.0...  
078946697X                                                NaN  
2070426769                                                NaN  
0789466953                                                NaN

In [ ]:
df_books_v2.loc['0751352497', 'Book-Author'] = 'Unknown'
df_books_v2.loc['1931696993', 'Publisher'] = 'NovelBooks Inc'
df_books_v2.loc['193169656X', 'Publisher'] = 'Novelbooks Inc'
df_books_v2.loc['9627982032', 'Book-Author'] = 'Larissa Anne Downes'

In [ ]:
df_books_v2[df_books_v2.isnull().any(axis=1)]

Book-Title Book-Author  \
ISBN                                                                        
078946697X  DK Readers: Creating the X-Men, How It All Beg...        2000   
2070426769  Peuple du ciel, suivi de 'Les Bergers\";Jean-M...        2003   
0789466953  DK Readers: Creating the X-Men, How Comic Book...        2000   

           Year-Of-Publication  \
ISBN                             
078946697X   DK Publishing Inc   
2070426769           Gallimard   
0789466953   DK Publishing Inc   

                                                    Publisher  \
ISBN                                                            
078946697X  http://images.amazon.com/images/P/078946697X.0...   
2070426769  http://images.amazon.com/images/P/2070426769.0...   
0789466953  http://images.amazon.com/images/P/0789466953.0...   

                                                  Image-URL-S  \
ISBN                                                            
078946697X  http://images.amazon.com/images/P/078946697X.0...   
2070426769  http://images.amazon.com/images/P/2070426769.0...   
0789466953  http://images.amazon.com/images/P/0789466953.0...   

                                                  Image-URL-M Image-URL-L  
ISBN                                                                       
078946697X  http://images.amazon.com/images/P/078946697X.0...         NaN  
2070426769  http://images.amazon.com/images/P/2070426769.0...         NaN  
0789466953  http://images.amazon.com/images/P/0789466953.0...         NaN

In [ ]:
# Tentar converter a coluna 'Year-Of-Publication' para números
df_books_v2['Year-Of-Publication'] = pd.to_numeric(df_books_v2['Year-Of-Publication'], errors='coerce')

# Verificar linhas onde a conversão falhou (valores NaN)
linhas_com_texto = df_books_v2[df_books_v2['Year-Of-Publication'].isna()]

print("Linhas com texto inválido na coluna 'Year-Of-Publication':")
print(linhas_com_texto)

Linhas com texto inválido na coluna 'Year-Of-Publication':
                                                   Book-Title Book-Author  \
ISBN                                                                        
078946697X  DK Readers: Creating the X-Men, How It All Beg...        2000   
2070426769  Peuple du ciel, suivi de 'Les Bergers\";Jean-M...        2003   
0789466953  DK Readers: Creating the X-Men, How Comic Book...        2000   

            Year-Of-Publication  \
ISBN                              
078946697X                  NaN   
2070426769                  NaN   
0789466953                  NaN   

                                                    Publisher  \
ISBN                                                            
078946697X  http://images.amazon.com/images/P/078946697X.0...   
2070426769  http://images.amazon.com/images/P/2070426769.0...   
0789466953  http://images.amazon.com/images/P/0789466953.0...   

                                                  Ima

In [ ]:
df_books_v2.loc['078946697X', 'Publisher'] = 'DK Publishing Inc'
df_books_v2.loc['2070426769', 'Publisher'] = 'Gallimard'
df_books_v2.loc['0789466953', 'Publisher'] = 'DK Publishing Inc'

df_books_v2.loc['078946697X', 'Image-URL-L'] = 'https://images.amazon.com/images/P/078946697X.0'
df_books_v2.loc['2070426769', 'Image-URL-L'] = 'https://images.amazon.com/images/P/2070426769.0'
df_books_v2.loc['0789466953', 'Image-URL-L'] = 'https://images.amazon.com/images/P/0789466953.0'

df_books_v2.loc['078946697X', 'Year-Of-Publication'] = '2000'
df_books_v2.loc['2070426769', 'Year-Of-Publication'] = '2003'
df_books_v2.loc['0789466953', 'Year-Of-Publication'] = '2000'

df_books_v2.loc['078946697X', 'Book-Author'] = "Cynthia O'Neill"
df_books_v2.loc['2070426769', 'Book-Author'] = 'Jean-Marie Gustave Le Clézio'
df_books_v2.loc['0789466953', 'Book-Author'] = "Cynthia O'Neill"

<ipython-input-8-9149389865ac>:9: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '2000' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df_books_v2.loc['078946697X', 'Year-Of-Publication'] = '2000'


In [ ]:
df_books_v2[df_books_v2.isnull().any(axis=1)]

Empty DataFrame
Columns: [Book-Title, Book-Author, Year-Of-Publication, Publisher, Image-URL-S, Image-URL-M, Image-URL-L]
Index: []

### Conversão da Coluna `Year-Of-Publication` para o Tipo Data e Ajustes nos Anos Inválidos

In [ ]:
df_books_v2['Year-Of-Publication'] = pd.to_numeric(df_books_v2['Year-Of-Publication'], errors='coerce')

df_books_v2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 271360 entries, 0195153448 to 0767409752
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Book-Title           271360 non-null  object 
 1   Book-Author          271360 non-null  object 
 2   Year-Of-Publication  271360 non-null  float64
 3   Publisher            271360 non-null  object 
 4   Image-URL-S          271360 non-null  object 
 5   Image-URL-M          271360 non-null  object 
 6   Image-URL-L          271360 non-null  object 
dtypes: float64(1), object(6)
memory usage: 24.6+ MB


In [ ]:
df_books_v2['Year-Of-Publication'].unique()

array([2002., 2001., 1991., 1999., 2000., 1993., 1996., 1988., 2004.,
       1998., 1994., 2003., 1997., 1983., 1979., 1995., 1982., 1985.,
       1992., 1986., 1978., 1980., 1952., 1987., 1990., 1981., 1989.,
       1984.,    0., 1968., 1961., 1958., 1974., 1976., 1971., 1977.,
       1975., 1965., 1941., 1970., 1962., 1973., 1972., 1960., 1966.,
       1920., 1956., 1959., 1953., 1951., 1942., 1963., 1964., 1969.,
       1954., 1950., 1967., 2005., 1957., 1940., 1937., 1955., 1946.,
       1936., 1930., 2011., 1925., 1948., 1943., 1947., 1945., 1923.,
       2020., 1939., 1926., 1938., 2030., 1911., 1904., 1949., 1932.,
       1928., 1929., 1927., 1931., 1914., 2050., 1934., 1910., 1933.,
       1902., 1924., 1921., 1900., 2038., 2026., 1944., 1917., 1901.,
       2010., 1908., 1906., 1935., 1806., 2021., 2012., 2006., 1909.,
       2008., 1378., 1919., 1922., 1897., 2024., 1376., 2037.])

In [ ]:
print(f"Quantidade de registros com anos iguais a zero: {df_books_v2[df_books_v2['Year-Of-Publication'] == 0].shape[0]}")
print(f"Quantidade de registros com anos menores que zero: {df_books_v2[df_books_v2['Year-Of-Publication'] < 0].shape[0]}")
print(f"Quantidade de registros com anos maiores que 2025: {df_books_v2[df_books_v2['Year-Of-Publication'] > 2025].shape[0]}")
print(df_books_v2[df_books_v2['Year-Of-Publication'] > 2025])

Quantidade de registros com anos iguais a zero: 4618
Quantidade de registros com anos menores que zero: 0
Quantidade de registros com anos maiores que 2025: 12
                                                   Book-Title  \
ISBN                                                            
0671746103  MY TEACHER FRIED MY BRAINS (RACK SIZE) (MY TEA...   
0671791990  MY TEACHER FLUNKED THE PLANET (RACK SIZE) (MY ...   
0870449842                                   Crossing America   
0140301690  Alice's Adventures in Wonderland and Through t...   
0140201092      Outline of European Architecture (Pelican S.)   
0394701658                       Three Plays of Eugene Oneill   
3442436893        Das gro��e B�¶se- M�¤dchen- Lesebuch.   
0870446924  Field Guide to the Birds of North America, 3rd...   
0671266500       FOREST PEOPLE (Touchstone Books (Hardcover))   
0684718022            In Our Time: Stories (Scribner Classic)   
068471809X                               To Have and Have No

In [ ]:
df_books_v2.loc['0671746103', 'Year-Of-Publication'] = 1991
df_books_v2.loc['0671791990', 'Year-Of-Publication'] = 2003
df_books_v2.loc['0870449842', 'Year-Of-Publication'] = 1999
df_books_v2.loc['0140301690', 'Year-Of-Publication'] = 1950
df_books_v2.loc['0140201092', 'Year-Of-Publication'] = 1950
df_books_v2.loc['0394701658', 'Year-Of-Publication'] = 1959
df_books_v2.loc['3442436893', 'Year-Of-Publication'] = 1996
df_books_v2.loc['0870446924', 'Year-Of-Publication'] = 1999
df_books_v2.loc['0671266500', 'Year-Of-Publication'] = 1961
df_books_v2.loc['0684718022', 'Year-Of-Publication'] = 1970
df_books_v2.loc['068471809X', 'Year-Of-Publication'] = 1937
df_books_v2.loc['0671740989', 'Year-Of-Publication'] = 1991

In [ ]:
df_books_v2[df_books_v2['Year-Of-Publication'] > 2025]

Empty DataFrame
Columns: [Book-Title, Book-Author, Year-Of-Publication, Publisher, Image-URL-S, Image-URL-M, Image-URL-L]
Index: []

In [ ]:
df_books_v2['Year-Of-Publication'].unique()

array([2002., 2001., 1991., 1999., 2000., 1993., 1996., 1988., 2004.,
       1998., 1994., 2003., 1997., 1983., 1979., 1995., 1982., 1985.,
       1992., 1986., 1978., 1980., 1952., 1987., 1990., 1981., 1989.,
       1984.,    0., 1968., 1961., 1958., 1974., 1976., 1971., 1977.,
       1975., 1965., 1941., 1970., 1962., 1973., 1972., 1960., 1966.,
       1920., 1956., 1959., 1953., 1951., 1942., 1963., 1964., 1969.,
       1954., 1950., 1967., 2005., 1957., 1940., 1937., 1955., 1946.,
       1936., 1930., 2011., 1925., 1948., 1943., 1947., 1945., 1923.,
       2020., 1939., 1926., 1938., 1911., 1904., 1949., 1932., 1928.,
       1929., 1927., 1931., 1914., 1934., 1910., 1933., 1902., 1924.,
       1921., 1900., 1944., 1917., 1901., 2010., 1908., 1906., 1935.,
       1806., 2021., 2012., 2006., 1909., 2008., 1378., 1919., 1922.,
       1897., 2024., 1376.])

In [ ]:
# Antes de converter para o tipo data, converter para tipo int
df_books_v2['Year-Of-Publication'] = df_books_v2['Year-Of-Publication'].astype('Int64')

df_books_v2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 271360 entries, 0195153448 to 0767409752
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   Book-Title           271360 non-null  object
 1   Book-Author          271360 non-null  object
 2   Year-Of-Publication  271360 non-null  Int64 
 3   Publisher            271360 non-null  object
 4   Image-URL-S          271360 non-null  object
 5   Image-URL-M          271360 non-null  object
 6   Image-URL-L          271360 non-null  object
dtypes: Int64(1), object(6)
memory usage: 24.9+ MB


In [ ]:
df_books_v2['Year-Of-Publication'] = pd.to_datetime(
    df_books_v2['Year-Of-Publication'].astype(str) + '-01-01',
    format='%Y-%m-%d',
    errors='coerce'
)

In [ ]:
df_books_v2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 271360 entries, 0195153448 to 0767409752
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   Book-Title           271360 non-null  object        
 1   Book-Author          271360 non-null  object        
 2   Year-Of-Publication  266740 non-null  datetime64[ns]
 3   Publisher            271360 non-null  object        
 4   Image-URL-S          271360 non-null  object        
 5   Image-URL-M          271360 non-null  object        
 6   Image-URL-L          271360 non-null  object        
dtypes: datetime64[ns](1), object(6)
memory usage: 24.6+ MB


In [ ]:
print(f"Quantidade de linhas com NaT no ano: {df_books_v2[df_books_v2.isnull().any(axis=1)].shape[0]}")

Quantidade de linhas com NaT no ano: 4620


## **`df_users`**

In [ ]:
print(f"Quantidade de registros na tabela users: {df_users_v2.shape[0]}")
print(f"Quantidade de registros com valores nulos em cada coluna: \n{df_users_v2.isna().sum()}")

Quantidade de registros na tabela users: 278858
Quantidade de registros com valores nulos em cada coluna: 
User-ID          0
Location         0
Age         110762
dtype: int64


In [ ]:
df_users_v2['Age'].unique()

array([ nan,  18.,  17.,  61.,  26.,  14.,  25.,  19.,  46.,  55.,  32.,
        24.,  20.,  34.,  23.,  51.,  31.,  21.,  44.,  30.,  57.,  43.,
        37.,  41.,  54.,  42.,  50.,  39.,  53.,  47.,  36.,  28.,  35.,
        13.,  58.,  49.,  38.,  45.,  62.,  63.,  27.,  33.,  29.,  66.,
        40.,  15.,  60.,   0.,  79.,  22.,  16.,  65.,  59.,  48.,  72.,
        56.,  67.,   1.,  80.,  52.,  69.,  71.,  73.,  78.,   9.,  64.,
       103., 104.,  12.,  74.,  75., 231.,   3.,  76.,  83.,  68., 119.,
        11.,  77.,   2.,  70.,  93.,   8.,   7.,   4.,  81., 114., 230.,
       239.,  10.,   5., 148., 151.,   6., 101., 201.,  96.,  84.,  82.,
        90., 123., 244., 133.,  91., 128.,  94.,  85., 141., 110.,  97.,
       219.,  86., 124.,  92., 175., 172., 209., 212., 237.,  87., 162.,
       100., 156., 136.,  95.,  89., 106.,  99., 108., 210.,  88., 199.,
       147., 168., 132., 159., 186., 152., 102., 116., 200., 115., 226.,
       137., 207., 229., 138., 109., 105., 228., 18

In [ ]:
usuarios_incoerentes = df_users_v2[(df_users_v2['Age'].isna()) |
                                   (df_users_v2['Age'] == 0) |
                                   (df_users_v2['Age'] > 110)]

usuarios_com_reviews = set(df_ratings_v2['User-ID'])

usuarios_incoerentes_sem_reviews = set(usuarios_incoerentes['User-ID']) - usuarios_com_reviews

df_users_filtrado = df_users_v2[~df_users_v2['User-ID'].isin(usuarios_incoerentes_sem_reviews)]

usuarios_restantes_sem_reviews = set(df_users_filtrado['User-ID']) - usuarios_com_reviews

df_users_v3 = df_users_filtrado[~df_users_filtrado['User-ID'].isin(usuarios_restantes_sem_reviews)]

print(f"Total de usuários antes da filtragem: {df_users_v2.shape[0]}")
print(f"Total de usuários após a filtragem: {df_users_v3.shape[0]}")
print(f"Usuários incoerentes sem reviews removidos: {len(usuarios_incoerentes_sem_reviews)}")
print(f"Usuários válidos sem reviews removidos: {len(usuarios_restantes_sem_reviews)}")
print(f"Total de usuários removidos: {df_users_v2.shape[0] - df_users_v3.shape[0]}")

Total de usuários antes da filtragem: 278858
Total de usuários após a filtragem: 105283
Usuários incoerentes sem reviews removidos: 67900
Usuários válidos sem reviews removidos: 105675
Total de usuários removidos: 173575


In [ ]:
df_users_v3['Age'].unique()

array([ 18.,  nan,  26.,  14.,  19.,  24.,  34.,  17.,  51.,  32.,  43.,
        44.,  37.,  20.,  41.,  42.,  57.,  49.,  47.,  38.,  45.,  35.,
        62.,  28.,  36.,  27.,  33.,  23.,  15.,  21.,  79.,  16.,  29.,
        25.,  53.,  59.,  63.,  22.,  40.,  48.,  72.,  31.,  58.,  50.,
        69.,  55.,  13.,  30.,  65.,   1.,  56.,  52.,  46.,   9.,  39.,
        71.,  74.,  54.,  61.,   0.,  75.,  60.,   3.,  64.,  66.,  83.,
       103.,  11.,  68.,   2.,  12.,  77.,  73.,  70., 104.,   8.,  76.,
       239.,  78.,  67.,  93., 151., 201.,  10.,  80., 244.,  82.,   7.,
       123., 128., 101.,  81., 141., 219.,  84.,   4.,  96., 172., 124.,
       133., 209., 212.,   5., 237., 156., 136.,  95.,  89.,   6.,  99.,
       100.,  86., 108., 199.,  92.,  85., 147.,  90.,  97., 168., 148.,
       132., 152., 114., 116., 226., 229., 138., 109., 105., 228., 223.,
       107., 157.,  91., 115., 146., 204., 118.,  94., 220., 200., 140.,
       127.,  98.])

In [ ]:
print(f"Quantidade de registros com idades acima que 110: {df_users_v3[df_users_v3['Age'] > 110].shape[0]}")
print(f"Quantidade de registros com idades iguais a zero: {df_users_v3[df_users_v3['Age'] == 0].shape[0]}")
print(f"Total de linhas que serão alteradas: {df_users_v3[df_users_v3['Age'] > 110].shape[0] + df_users_v3[df_users_v3['Age'] == 0].shape[0]}")

Quantidade de registros com idades acima que 110: 50
Quantidade de registros com idades iguais a zero: 148
Total de linhas que serão alteradas: 198


In [ ]:
df_users_v3.loc[(df_users_v3['Age'] > 110) | (df_users_v3['Age'] == 0), 'Age'] = np.nan

In [ ]:
df_users_v3['Age'].unique()

array([ 18.,  nan,  26.,  14.,  19.,  24.,  34.,  17.,  51.,  32.,  43.,
        44.,  37.,  20.,  41.,  42.,  57.,  49.,  47.,  38.,  45.,  35.,
        62.,  28.,  36.,  27.,  33.,  23.,  15.,  21.,  79.,  16.,  29.,
        25.,  53.,  59.,  63.,  22.,  40.,  48.,  72.,  31.,  58.,  50.,
        69.,  55.,  13.,  30.,  65.,   1.,  56.,  52.,  46.,   9.,  39.,
        71.,  74.,  54.,  61.,  75.,  60.,   3.,  64.,  66.,  83., 103.,
        11.,  68.,   2.,  12.,  77.,  73.,  70., 104.,   8.,  76.,  78.,
        67.,  93.,  10.,  80.,  82.,   7., 101.,  81.,  84.,   4.,  96.,
         5.,  95.,  89.,   6.,  99., 100.,  86., 108.,  92.,  85.,  90.,
        97., 109., 105., 107.,  91.,  94.,  98.])

In [ ]:
df_users_v3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 105283 entries, 1 to 278853
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   105283 non-null  int64  
 1   Location  105283 non-null  object 
 2   Age       61909 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 3.2+ MB


In [ ]:
df_users_v3['Age'] = df_users_v3['Age'].astype('Int64')

<ipython-input-28-f1a644298018>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_users_v3['Age'] = df_users_v3['Age'].astype('Int64')


In [ ]:
df_users_v3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 105283 entries, 1 to 278853
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   User-ID   105283 non-null  int64 
 1   Location  105283 non-null  object
 2   Age       61909 non-null   Int64 
dtypes: Int64(1), int64(1), object(1)
memory usage: 3.3+ MB


In [ ]:
df_users_v2 = df_users_v3

df_users_v2.shape[0]

105283

In [ ]:
localizacoes_codificadas = df_users_v2[df_users_v2['Location'].str.contains('&#', na=False)]
print(localizacoes_codificadas)

        User-ID                                           Location   Age
9310       9311                    &#321;ód&#378;, lodzkie, poland    17
10661     10662                  nanchong, &#22235;&#24029;, china  <NA>
26569     26570       b&#322;a&#380;owa, woj. podkarpackie, poland    15
33775     33776               £ód&#378;, woj. &#322;ódzkie, poland    32
34899     34900                        &#321;ód&#378;, n/a, poland    67
35955     35956                         bodrum, mu&#287;la, turkey    59
55899     55900               &#321;ód&#378;, &#322;ódzkie, poland    29
71866     71867                            ayd&#305;n, n/a, turkey    19
100374   100375                kocaeli, &#304;nanç  lisesi, turkey    15
104295   104296          sosnowiec, woj. &#346;l&#261;skie, poland    29
124783   124784                     &#24034;&#28246;, anhui, china  <NA>
159302   159303                  bia&#322;ystok, podlaskie, poland    38
180842   180843                     &#269;akovec, c

In [ ]:
# Decodificar as localizações codificadas
df_users_v2['Location'] = df_users_v2['Location'].apply(lambda x: html.unescape(x) if isinstance(x, str) and '&#' in x else x)

# Padronizar as localizações, transformando para minúsculas e removendo espaços extras
df_users_v2['Location'] = df_users_v2['Location'].str.lower().str.strip()

<ipython-input-32-2ab0c0f6057b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_users_v2['Location'] = df_users_v2['Location'].apply(lambda x: html.unescape(x) if isinstance(x, str) and '&#' in x else x)
<ipython-input-32-2ab0c0f6057b>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_users_v2['Location'] = df_users_v2['Location'].str.lower().str.strip()


In [ ]:
print(df_users_v2[df_users_v2['User-ID'] == 189961])

        User-ID       Location  Age
189960   189961  贵港, 广西, china   18


In [ ]:
# Resetar a coluna 'index' do df_books_v2
df_books_v2_reset = df_books_v2.reset_index()

# Renomear a coluna 'index' para 'ISBN' no DataFrame de livros
df_books_v2.rename(columns={'index': 'ISBN'}, inplace=True)

In [ ]:
from google.colab import files

# Salvar os DataFrames em arquivos CSV
df_users_v2.to_csv("df_users_v2.csv", index=False)
df_books_v2_reset.to_csv("df_books_v2.csv", index=False)
df_ratings_v2.to_csv("df_ratings_v2.csv", index=False)

# Treinamento do Modelo

## Importação das Bibliotecas

In [ ]:
pip install scikit-surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.4/154.4 kB 3.4 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.4-cp311-cp311-linux_x86_64.whl size=2505202 sha256=4c89e5eac21e36d850679a93591927c3cf3a27d4df956384de1fe4aa9113d2d0
  Stored in directory: /root/.cache/pip/wheels/2a/8f/6e/7e2899163e2d85d8266daab4aa1cdabec7a6c56f83c015b5af
Successfully built scikit-surprise


In [1]:
pip install numpy==1.23.5

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from scipy.sparse import csr_matrix
from surprise import SVD, Dataset, Reader
from surprise.model_selection import cross_validate

## Carregamento de Dados

In [3]:
df_books_v2 = pd.read_csv('df_books_v2.csv')
df_ratings_v2 = pd.read_csv('df_ratings_v2.csv')
df_users_v2 = pd.read_csv('df_users_v2.csv')

In [4]:
print(f"Colunas do dataframe books: {df_books_v2.columns}\n")
print(f"Colunas do dataframe ratings: {df_ratings_v2.columns}\n")
print(f"Colunas do dataframe users: {df_users_v2.columns}\n")

Colunas do dataframe books: Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

Colunas do dataframe ratings: Index(['User-ID', 'ISBN', 'Book-Rating'], dtype='object')

Colunas do dataframe users: Index(['User-ID', 'Location', 'Age'], dtype='object')



## Filtragem de Usuários por Idade e Localização (País)

### Aplicação de filtros adicionais com avaliações de livros

Este código realiza a filtragem dos dados de usuários, considerando informações de idade e país.

Ele define faixas etárias relevantes e extrai o país da localização para segmentar os usuários.

 As etapas de filtragem incluem:

* **Filtragem por idade**: Remove usuários com idade abaixo de um limite mínimo.
Filtragem por país: Seleciona usuários de países específicos considerados relevantes.
* **Filtragem por atividade do usuário**: Mantém apenas usuários com um número mínimo de avaliações.
* **Filtragem por popularidade do livro:** Inclui apenas livros com um número mínimo de avaliações.

Essas etapas garantem que o sistema de recomendação trabalhe com dados de alta qualidade, focando em usuários ativos e livros populares em regiões relevantes.

In [5]:
# Parâmetros de filtragem
idade_minima = 20  # Define a idade mínima para considerar um usuário relevante (20 anos)
paises_relevantes = ['usa', 'canada', 'uk']  # Lista de países relevantes para o sistema de recomendação

# Tratar valores ausentes na coluna 'Age'
df_users_v2['Age'] = df_users_v2['Age'].fillna(df_users_v2['Age'].mean()) # Preenche valores ausentes na coluna 'Age' com a média de idade

# Filtrar usuários por idade
usuarios_idade_filtrada = df_users_v2[df_users_v2['Age'] >= idade_minima]['User-ID'].unique() # Obtém IDs de usuários com idade igual ou superior à idade mínima
avaliacoes_filtradas = df_ratings_v2[df_ratings_v2['User-ID'].isin(usuarios_idade_filtrada)] # Filtra as avaliações para incluir apenas usuários com idade filtrada

# Filtrar usuários por país (extrair país)
def extrair_pais(location):
    try:
        pais = location.split(',')[-1].strip().lower() # Extrai o país da string de localização, remove espaços e converte para minúsculo
        if pais:
            return pais
        else:
            return None # Retorna None se o país estiver vazio
    except:
        return None # Retorna None em caso de erro ao extrair o país

df_users_v2['Pais'] = df_users_v2['Location'].apply(extrair_pais) # Aplica a função para extrair o país da coluna 'Location' e cria a coluna 'Pais'
usuarios_pais_filtrada = df_users_v2[df_users_v2['Pais'].isin(paises_relevantes)]['User-ID'].unique() # Obtém IDs de usuários dos países relevantes
avaliacoes_filtradas = avaliacoes_filtradas[avaliacoes_filtradas['User-ID'].isin(usuarios_pais_filtrada)] # Filtra as avaliações para incluir apenas usuários dos países filtrados

# Filtrar usuários com um número mínimo de avaliações
minimo_avaliacoes_usuario = 20  # Define o número mínimo de avaliações que um usuário deve ter para ser considerado ativo
contagem_avaliacoes_usuario = avaliacoes_filtradas['User-ID'].value_counts()  # Conta quantas avaliações cada usuário fez
usuarios_ativos = contagem_avaliacoes_usuario[contagem_avaliacoes_usuario >= minimo_avaliacoes_usuario].index  # Obtém os IDs dos usuários ativos
avaliacoes_filtradas = avaliacoes_filtradas[avaliacoes_filtradas['User-ID'].isin(usuarios_ativos)]  # Filtra as avaliações para incluir apenas usuários ativos

# Filtrar livros com um número mínimo de avaliações
minimo_avaliacoes_livro = 20 # Define o número mínimo de avaliações que um livro deve ter para ser considerado popular
contagem_avaliacoes_livro = df_ratings_v2['ISBN'].value_counts() # Conta quantas avaliações cada livro recebeu
livros_populares = contagem_avaliacoes_livro[contagem_avaliacoes_livro >= minimo_avaliacoes_livro].index # Obtém os ISBNs dos livros populares
avaliacoes_filtradas = df_ratings_v2[df_ratings_v2['ISBN'].isin(livros_populares)] # Filtra as avaliações para incluir apenas livros populares

## Criação da Matriz de Utilidade
### Conversão para otimização da memória

A matriz de utilidade é uma tabela que mostra as interações entre usuários e itens, onde cada célula contém a avaliação de um usuário para um livro. Ela foi escolhida porque é muito usada em sistemas de recomendação, ajudando a entender as preferências dos usuários e prever quais livros podem interessá-los no futuro, com base nas avaliações passadas.

Este código tem como objetivo criar uma matriz de utilidade a partir dos dados de avaliações filtradas. A matriz de utilidade é uma representação crucial para sistemas de recomendação colaborativos, onde as linhas representam usuários, as colunas representam livros (ISBNs) e os valores representam as avaliações dos usuários para os livros.

A matriz de utilidade foi então convertida para o formato *csr_matrix* (*Compressed Sparse Row matrix*). Este formato é otimizado para matrizes esparsas, que são comuns em sistemas de recomendação onde a maioria dos usuários avaliou apenas uma pequena fração dos livros disponíveis. A conversão para *csr_matrix* reduz o uso de memória e acelera cálculos, tornando o sistema de recomendação mais eficiente.

In [6]:
# Cria a matriz de utilidade a partir das avaliações filtradas, preenchendo valores ausentes com 0
matriz_utilidade = avaliacoes_filtradas.pivot_table(index='User-ID', columns='ISBN', values='Book-Rating').fillna(0)

# Converte a matriz de utilidade para o formato csr_matrix para otimização de memória e cálculos
matriz_utilidade_csr = csr_matrix(matriz_utilidade.values)

print(f"Tamanho da matriz de utilidade filtrada: {matriz_utilidade_csr.shape}")

Tamanho da matriz de utilidade filtrada: (57378, 7490)


## Treinamento e Avaliação de um Modelo de Recomendação SVD

### Utilizando a Biblioteca Surprise para Sistemas de Recomendação Colaborativos

Este código demonstra o processo de treinamento e avaliação de um modelo de recomendação utilizando o algoritmo *SVD* (*Singular Value Decomposition*) da biblioteca `Surprise`.

O *SVD* é um algoritmo popular para sistemas de recomendação colaborativos, que decompõe a matriz de utilidade em fatores latentes, permitindo prever as avaliações que um usuário daria a um livro.

O código segue as seguintes etapas:

* **Preparação dos dados**: Cria um Reader para definir o formato dos dados de avaliação e carrega os dados filtrados em um objeto Dataset da Surprise.
* **Treinamento do modelo**: Cria uma instância do modelo SVD e o treina com os dados de avaliação.
* **Avaliação do modelo**: Utiliza validação cruzada para avaliar o desempenho do modelo em diferentes partições dos dados.
* **Treinamento final**: Treina o modelo com todos os dados disponíveis para maximizar o aprendizado.


In [7]:
# Criar o Reader para indicar o formato dos dados
reader = Reader(rating_scale=(1, 10)) # Cria um objeto Reader para definir a escala de avaliação (1 a 10)

avaliacoes_filtradas = df_ratings_v2 # Atribui o DataFrame de avaliações filtradas à variável avaliacoes_filtradas

# Criar o dataset a partir do DataFrame filtrado
avaliacoes_dataset = Dataset.load_from_df(avaliacoes_filtradas[['User-ID', 'ISBN', 'Book-Rating']], reader) # Cria um objeto Dataset a partir do DataFrame, especificando as colunas relevantes e o Reader

# Criar o modelo SVD
modelo_svd = SVD() # Cria uma instância do modelo SVD

# Avaliação do modelo com validação cruzada
resultados = cross_validate(modelo_svd, avaliacoes_dataset, cv=5, verbose=True) # Avalia o modelo usando validação cruzada com 5 folds e imprime os resultados

# Treinar o modelo com todos os dados
trainset = avaliacoes_dataset.build_full_trainset() # Cria um objeto Trainset com todos os dados de avaliação
modelo_svd.fit(trainset) # Treina o modelo SVD com o conjunto de treinamento completo

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    3.4956  3.4949  3.4990  3.4979  3.4979  3.4971  0.0015  
MAE (testset)     2.9235  2.9250  2.9318  2.9283  2.9258  2.9269  0.0029  
Fit time          27.00   27.88   27.39   27.44   27.94   27.53   0.35    
Test time         3.17    2.39    3.14    2.72    3.22    2.93    0.32    


### Gerando Recomendações para um Usuário Específico

Este código implementa uma função `recomendar_livros_svd` que gera recomendações de livros personalizadas para um usuário específico, utilizando o modelo *SVD* treinado anteriormente.

A função identifica os livros que o usuário ainda não avaliou, prevê as avaliações que o usuário daria a esses livros e, em seguida, recomenda os livros com as previsões de avaliação mais altas.

In [8]:
def recomendar_livros_svd(id_usuario, df_books_v2, n_recomendacoes=3):
    livros_nao_avaliados = df_books_v2[~df_books_v2['ISBN'].isin(avaliacoes_filtradas[avaliacoes_filtradas['User-ID'] == id_usuario]['ISBN'])]

    # Prever notas para os livros não avaliados
    predicoes = [(isbn, modelo_svd.predict(id_usuario, isbn).est) for isbn in livros_nao_avaliados['ISBN']]

    # Ordenar pelos livros com melhor previsão
    predicoes.sort(key=lambda x: x[1], reverse=True)

    # Selecionar os melhores N livros
    recomendacoes = [df_books_v2[df_books_v2['ISBN'] == isbn]['Book-Title'].values[0] for isbn, _ in predicoes[:n_recomendacoes]]

    return recomendacoes

id_usuario = 99
livros_recomendados = recomendar_livros_svd(id_usuario, df_books_v2)

print(f"Livros recomendados para o usuário {id_usuario}:")
for livro in livros_recomendados:
    print(f"- {livro}")

Livros recomendados para o usuário 99:
- Free
- The WEDDING
- Good in Bed


# Interpretação das métricas de validação cruzada com 5 divisões (folds)

* **RMSE (Root Mean Squared Error)**
  * Média (Mean): 3.4971
  * Desvio padrão (Std): 0.0015

O RMSE é uma medida de quão distantes estão as previsões dos valores reais. Quanto menor for o valor de RMSE, melhor. Neste caso, o modelo SVD tem uma boa precisão, com um RMSE consistente entre os folds.

* **MAE (Mean Absolute Error)**
  * Média (Mean): 2.9269
  * Desvio padrão (Std): 0.0029

O MAE mede a média das diferenças absolutas entre as previsões e os valores reais. Menor MAE também indica que as previsões estão mais próximas dos valores reais. O valor de 2.93 sugere uma precisão razoável no modelo, mas ainda há espaço para melhorias.

* **Fit time (Tempo de treinamento)**
  * Média (Mean): 27.53 segundos
  * Desvio padrão (Std): 0.35 segundos

O tempo de treinamento é o tempo necessário para o modelo aprender a partir dos dados. Com base nesses resultados, o tempo de treinamento é estável entre os folds (~27.5s), indicando que a complexidade do treinamento não varia muito

* **Test time (Tempo de teste)**
  * Média (Mean): 2.93 segundos
  * Desvio padrão (Std): 0.32 segundos

O tempo de teste é o tempo necessário para fazer as previsões no conjunto de teste. O modelo leva em média 3 segundos para testar.

# Conclusão
O modelo SVD tem um desempenho bom em termos de RMSE e MAE, com um erro médio relativamente baixo. O tempo de ajuste e teste também são razoáveis, considerando que o modelo foi validado em 5 folds.

Os próximos passos consistem em ajustar os parâmetros do modelo SVD, como n_factors, lr_all, ou reg_all, para melhorar a precisão.